### DISCUSSION TITLE
State of Tai Tham support?

### DISCUSSION THREAD

**Discussion by Eli-Zaretskii** [harfbuzz/harfbuzz#3430] [Q&A] [answered]
Can someone please describe the state of support of Tai Tham shaping in latest HarfBuzz?

The last discussions here seem to be from 3 years ago, and in the current sources I see Tai Tham mentioned only in the USE-related files, which could mean this script is only supported on MS-Windows with the USE backend compiled in?

So I'd appreciate if someone could describe the current state, and maybe also point to the source of the data and algorithms for the implemented shaping.  Thanks in advance!

**Comment by dscorbett** [comment, marked as answer]
Unicode does not specify how to encode Tai Tham text. [L2/19-365](https://www.unicode.org/L2/L2019/19365-tai-tham-structure.pdf) is one proposal of how to fix the problem. It is not the only possible solution. In particular, some people might prefer a visually ambiguous, phonetic encoding. In any case, progress seems to have stalled.

OpenType specifies that Tai Tham should use the [USE](https://docs.microsoft.com/en-us/typography/script-development/use), albeit with a caveat: “Tai Tham support is currently limited. Additional encoding work is required for full text representation to be possible.” It overrides the syllabic categories of [U+1A7B, U+1A7C, and U+1A7F](https://github.com/microsoft/font-tools/blob/c355d0419429fca2f1a22d88d0afd35822e4c0b0/USE/IndicSyllabicCategory-Additional.txt#L57-L59), but otherwise doesn’t do anything special. Different candidate Tai Tham encodings are therefore subject to various levels of unwanted dotted circle insertion.

HarfBuzz implements USE for Tai Tham mostly as specified, inheriting the spec’s limitations, but with a tweak to U+1A60 TAI THAM SIGN SAKOT. After all vowel signs and vowel modifiers, there can be any number of <sakot, base>. This is to support bases subjoined to post-base vowel signs. This is experimental and may be reverted once Unicode gets its act together.

Currently, the safest way to make a Tai Tham font is to ignore USE and use the default shaper instead. You have to implement your own vowel reordering but at least there won’t be any extraneous dotted circles. This allows every Tai Tham font designer the freedom to invent their own encoding. Hopefully, Unicode will one day provide a useful specification and 'lana' will become standard.

**Comment by Eli-Zaretskii** [comment]
> Date: Tue, 15 Feb 2022 08:16:30 -0800
> From: David Corbett ***@***.***>
> Cc: Eli-Zaretskii ***@***.***>, 
>  Author ***@***.***>
> 
> Unicode does not specify how to encode Tai Tham text. L2/19-365 is one proposal of how to fix the problem.
> It is not the only possible solution. In particular, some people might prefer a visually ambiguous, phonetic
> encoding. In any case, progress seems to have stalled.

Can you please clarify what you mean by "encoding" here?  The Unicode
Standard has the Tai Tham block, as I'm sure you know, so by "encode"
you probably didn't mean the codepoints of the Tai Tham script
characters.

(Please forgive me the possibly silly questions: my knowledge of
text shaping is quite rudimentary.)

> OpenType specifies that Tai Tham should use the USE,

By "use the USE" you mean use the rules defined for the USE, is that
right?

> Currently, the safest way to make a Tai Tham font is to ignore USE and use the default shaper instead. You
> have to implement your own vowel reordering but at least there won’t be any extraneous dotted circles. This
> allows every Tai Tham font designer the freedom to invent their own encoding. Hopefully, Unicode will one
> day provide a useful specification and 'lana' will become standard.

Thanks, but the purpose of my question was not to work on a font, but
to add Tai Tham support for Emacs that uses HarfBuzz as its shaping
engine.

I think I have my answers, thank you very much.

**Reply by dscorbett** [reply]
> Can you please clarify what you mean by "encoding" here?

By encoding a certain piece of text, I mean creating a code point sequence that represents that text unambiguously and interoperably. Unicode is responsible not just for providing code points for a script’s characters but also for explaining how to use those code points to represent text. The code points themselves aren’t very useful if it’s not clear how to combine them into strings. (Similar problems exist for Duployan, Mongolian, SignWriting, and probably more.)

> By "use the USE" you mean use the rules defined for the USE, is that
right?

Yes.
